storages/backends/gcloud.py

```diff
@@ -0,0 +1,241 @@
+import mimetypes
+from tempfile import SpooledTemporaryFile
+
+from django.core.exceptions import ImproperlyConfigured
+from django.core.files.base import File
+from django.utils.encoding import force_bytes, force_text, smart_str
+from storages.compat import Storage
+from storages.utils import setting
+
+try:
+    from google.cloud.storage.client import Client
+    from google.cloud.storage.bucket import Bucket
+    from google.cloud.storage.blob import Blob
+    from google.cloud.exceptions import NotFound
+except ImportError:
+    raise ImproperlyConfigured("Could not load Google Storage bindings.\n"
+                               "See https://github.com/GoogleCloudPlatform/gcloud-python")
+
+
+class GoogleCloudFile(File):
+    def __init__(self, name, mode, storage, buffer_size=None):
+        self.name = name
+        self._mode = mode
+        self._storage = storage
+        self.blob = Blob(self.name, storage.bucket)
+        self._file = None
+        self._is_dirty = False
+
+    @property
+    def size(self):
+        return self.blob.size
+
+    def _get_file(self):
+        if self._file is None:
+            self._file = SpooledTemporaryFile(
+                max_size=self._storage.max_memory_size,
+                suffix=".GSStorageFile",
+                dir=setting("FILE_UPLOAD_TEMP_DIR", None)
+            )
+            if 'r' in self._mode:
+                self._is_dirty = False
+                self.blob.download_to_file(self._file)
+                self._file.seek(0)
+        return self._file
+
+    def _set_file(self, value):
+        self._file = value
+
+    file = property(_get_file, _set_file)
+
+    def read(self, *args, **kwargs):
+        if 'r' not in self._mode:
+            raise AttributeError("File was not opened in read mode.")
+        return super(GoogleCloudFile, self).read(*args, **kwargs)
+
+    def write(self, content, *args, **kwargs):
+        if 'w' not in self._mode:
+            raise AttributeError("File was not opened in write mode.")
+        self._is_dirty = True
+        return super(GoogleCloudFile, self).write(force_bytes(content), *args, **kwargs)
+
+    def close(self):
+        if self._file is not None:
+            if self._is_dirty:
+                self.file.seek(0)
+                content_type, _ = mimetypes.guess_type(self.name)
+                content_type = getattr(self.file, 'content_type', content_type)
+                size = getattr(self.file, 'size')
+                self.blob.upload_from_file(self.file, content_type=content_type, size=size)
+            self._file.close()
+            self._file = None
+
+
+class GoogleCloudStorage(Storage):
```

Comment:
@eirsyl The existing backends all use the `@deconstructible` decorator here but I'm not sure why. Commit comments suggest it has something to do with migrations but I'm able to migrate to a `FileField` using this backend for storage with this code as is. Therefore I'm inclined to leave the decorator out and if someone needs it, they can add it. What do you think?